# MBA Specialization Classification

![](mba1.png)

### Contents
1. Abstract
2. Dataset
3. Importing required libraries
4. Data pre-processing
5. Finding out the correlation between the attributes
6. Spliting the dataset for model creation
7. Classification Models -
    - Logistic Regression
    - Decision Tree Classifier
    - Random forest classifier
    - Gaussian NB
    - K-Nearest Neighbouring
    - Support Vector Machine (SVM)
    * Gradient Boosting
    - MLP Classifier
    - Stochastic Gradient Descent (SGD)
    - Linear Discriminant Analysis (LDA)
8. Model Comparison
9. Conclusion

### Abstract
An MBA in sales and marketing will be involved in the core sales process of the organisations and will be expected to connect with clients, generating value and revenue. They will be responsible for end to end Business cycle management, right from engagement with the client to the signing of contracts. They will be responsible for new business development involving prospect development, understanding business and technology scenarios and needs, identifying opportunities, solution offerings and business cases to the client & involvement in preparation of proposals. They will also be responsible for relationship building and account management. The prime job role will be to manage sales with the sole focus of maximizing revenues in a timely, reliable and consistent basis.

In the present day many organisations are expecting MBA ops graduates to be well versed in mobile technologies. An MBA in operations are also expected to take on challenges in logistics, supply chain design and supply chain management. They are made to handle various strategic issues including determining project management methods and implementing the structure of information technology networks and other operational issues include inventory management, work-in-process levels, acquisition of raw materials, quality control, handling of materials and maintenance policies etc.

### Goal
The goal of this project is to find out the MBA students who will be having good scores in their MBA career based on past activities.

### Dataset
The dataset is collected from Kaggle website. Here is the link for the website : https://www.kaggle.com/balakrishcodes/others?select=MBA_ADMISSIONS.csv

### Importing required libraries and dataset

In [63]:
# Importing Libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')
plt.style.use('ggplot')

In [64]:
data = pd.read_csv('MBA_ADMISSIONS.csv')
data.head()

pre_score  Age_in_years  Percentage_in_10_Class  Percentage_in_12_Class  \
0  75.000000            22                    71.0                    74.8   
1  71.666667            25                    77.6                    82.6   
2  76.666667            26                    93.2                    83.8   
3  66.666667            22                    91.2                    80.0   
4  71.666667            24                    79.8                    61.6   

   Percentage_in_Under_Graduate  percentage_MBA  post_score  Gender  \
0                         72.00           61.00   83.333333    Male   
1                         76.90           66.85   76.666667    Male   
2                         77.00           74.97   75.000000  Female   
3                         67.00           68.30   60.000000    Male   
4                         60.33           69.28   76.666667  Female   

          STATE Previous_Degree Marital_status Place_you_belong_to  \
0  Central Zone     Engineering         Single               Urban   
1  Central Zone     Engineering         Single          Semi Urban   
2  Central Zone     Engineering         Single               Urban   
3  Central Zone        Commerce         Single          Semi Urban   
4  Central Zone     Engineering         Single               Urban   

  perceived#Job#Skill Specialization  
0     prefered skills      Marketing  
1     prefered skills            LOS  
2      desired skills        Finance  
3     prefered skills        Finance  
4     prefered skills        Finance

**Data Pre-processing**

In [95]:
df = data.astype(int, errors='ignore')

In [96]:
df.head()

pre_score  Age_in_years  Percentage_in_10_Class  Percentage_in_12_Class  \
0         75            22                      71                      74   
1         71            25                      77                      82   
2         76            26                      93                      83   
3         66            22                      91                      80   
4         71            24                      79                      61   

   Percentage_in_Under_Graduate  percentage_MBA  post_score  Gender  \
0                            72              61          83    Male   
1                            76              66          76    Male   
2                            77              74          75  Female   
3                            67              68          60    Male   
4                            60              69          76  Female   

          STATE Previous_Degree Marital_status Place_you_belong_to  \
0  Central Zone     Engineering         Single               Urban   
1  Central Zone     Engineering         Single          Semi Urban   
2  Central Zone     Engineering         Single               Urban   
3  Central Zone        Commerce         Single          Semi Urban   
4  Central Zone     Engineering         Single               Urban   

  perceived#Job#Skill Specialization  
0     prefered skills      Marketing  
1     prefered skills            LOS  
2      desired skills        Finance  
3     prefered skills        Finance  
4     prefered skills        Finance

**Finding out the correlation between the attributes of the dataset**

In [97]:
df.corr()

pre_score  Age_in_years  Percentage_in_10_Class  \
pre_score                      1.000000      0.250699               -0.140844   
Age_in_years                   0.250699      1.000000                0.038332   
Percentage_in_10_Class        -0.140844      0.038332                1.000000   
Percentage_in_12_Class        -0.191213     -0.207313                0.441987   
Percentage_in_Under_Graduate   0.009590     -0.141367                0.410374   
percentage_MBA                 0.053699      0.296556                0.490386   
post_score                     0.265957     -0.036615               -0.075099   

                              Percentage_in_12_Class  \
pre_score                                  -0.191213   
Age_in_years                               -0.207313   
Percentage_in_10_Class                      0.441987   
Percentage_in_12_Class                      1.000000   
Percentage_in_Under_Graduate                0.418453   
percentage_MBA                              0.286299   
post_score                                  0.052577   

                              Percentage_in_Under_Graduate  percentage_MBA  \
pre_score                                         0.009590        0.053699   
Age_in_years                                     -0.141367        0.296556   
Percentage_in_10_Class                            0.410374        0.490386   
Percentage_in_12_Class                            0.418453        0.286299   
Percentage_in_Under_Graduate                      1.000000        0.363450   
percentage_MBA                                    0.363450        1.000000   
post_score                                        0.161584        0.067904   

                              post_score  
pre_score                       0.265957  
Age_in_years                   -0.036615  
Percentage_in_10_Class         -0.075099  
Percentage_in_12_Class          0.052577  
Percentage_in_Under_Graduate    0.161584  
percentage_MBA                  0.067904  
post_score                      1.000000

**Creating the dataset for training the model**

In [107]:
x = df[['Percentage_in_10_Class','Percentage_in_Under_Graduate','Percentage_in_12_Class']]

In [108]:
y = df['percentage_MBA']

<a id="train-test-split"></a>
**Training and Testing Dataset Spliting using the `train_test_split`**
  
  * Immporting the library from the sklearn.model_selection
  * Split the dataset into 80:20 ratio
  * x_train1 and y_train1 are the trainning datasets
  * x_test1 and y_test1 are the testing datasets
  * After the spliting of the datasets the model is ready to be prepared!

In [109]:
from sklearn.model_selection import train_test_split

In [110]:
x_train1, x_test1, y_train1, y_test1 = train_test_split(x,y, test_size = 0.2)

In [111]:
from sklearn import preprocessing
from sklearn import utils

In [112]:
x_train1.shape

(378, 3)

## Classification Algorithms
Classification is a task that requires the use of machine learning algorithms that learn how to assign a class label to examples from the problem domain. An easy to understand example is classifying emails as “spam” or “not spam.”

There are many different types of classification tasks that you may encounter in machine learning and specialized approaches to modeling that may be used for each.

Here I am going to use 10 Classification algorithms, based on these the models will be trained and then evaluated using the accuracy scores.

**The following models that we are going to use -**
  * **Logistic Regression** : Logistic regression is a statistical model that in its basic form uses a logistic function to model a binary dependent variable, although many more complex extensions exist. In regression analysis, logistic regression (or logit regression) is estimating the parameters of a logistic model (a form of binary regression).
  
  
  * **Decision Tree Classifier** : Decision Tree is a Supervised learning technique that can be used for both classification and Regression problems, but mostly it is preferred for solving Classification problems. It is a tree-structured classifier, where internal nodes represent the features of a dataset, branches represent the decision rules and each leaf node represents the outcome.
  
  
  * **Random Forest Classifier** : Random Forest is a popular machine learning algorithm that belongs to the supervised learning technique. It can be used for both Classification and Regression problems in ML. It is based on the concept of ensemble learning, which is a process of combining multiple classifiers to solve a complex problem and to improve the performance of the model.
  
  
  * **Gausian NB** : This method is expected to be called several times consecutively on different chunks of a dataset so as to implement out-of-core or online learning. This is especially useful when the whole dataset is too big to fit in memory at once. This method has some performance and numerical stability overhead, hence it is better to call partial_fit on chunks of data that are as large as possible (as long as fitting in the memory budget) to hide the overhead.
  
  
  * **KNN algorithm** : K-Nearest Neighbour is one of the simplest Machine Learning algorithms based on Supervised Learning technique. K-NN algorithm assumes the similarity between the new case/data and available cases and put the new case into the category that is most similar to the available categories. K-NN algorithm stores all the available data and classifies a new data point based on the similarity. This means when new data appears then it can be easily classified into a well suite category by using K- NN algorithm.
  
  
  * **Support Vector Machine Algorithm** : Support Vector Machine or SVM is one of the most popular Supervised Learning algorithms, which is used for Classification as well as Regression problems. However, primarily, it is used for Classification problems in Machine Learning. The goal of the SVM algorithm is to create the best line or decision boundary that can segregate n-dimensional space into classes so that we can easily put the new data point in the correct category in the future. This best decision boundary is called a hyperplane.
  

* **Stochastic Gradient Descent Classifier** : Stochastic Gradient Descent (SGD) is a simple yet efficient optimization algorithm used to find the values of parameters/coefficients of functions that minimize a cost function. In other words, it is used for discriminative learning of linear classifiers under convex loss functions such as SVM and Logistic regression.


* **Linear Discriminant Analysis (LDA)** : Linear Discriminant Analysis (LDA) is a dimensionality reduction technique. As the name implies dimensionality reduction techniques reduce the number of dimensions (i.e. variables) in a dataset while retaining as much information as possible.


* **Gradient Boosting** : Gradient boosting is a machine learning technique for regression, classification and other tasks, which produces a prediction model in the form of an ensemble of weak prediction models, typically decision trees.


* **MLP Classifier** : Multi-layer perceptrons (MLP) make powerful classifiers that may provide superior performance compared with other classifiers, but are often criticized for the number of free parameters. Parameter selection for optimal performance is performed using measures that correlate well with generalisation error.


 
 
We are going to use these ten algorithms and based on the scores of the models the most fitted algorithm will be set! Now let's check out the algorithms.

In [113]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from time import time
from sklearn.metrics import f1_score
from os import path, makedirs, walk
from joblib import dump, load
import json

### Logistic Regression 

Logistic regression is a statistical model that in its basic form uses a logistic function to model a binary dependent variable, although many more complex extensions exist. In regression analysis, logistic regression (or logit regression) is estimating the parameters of a logistic model (a form of binary regression).

In [114]:
from sklearn.linear_model import LogisticRegression
logReg = LogisticRegression(max_iter = 1000)
logReg.fit(x_train1, y_train1)

LogisticRegression(max_iter=1000)

In [115]:
logReg.score(x_test1, y_test1)

0.23157894736842105

### Decision Tree Classifier Algorithm

Decision Tree is a Supervised learning technique that can be used for both classification and Regression problems, but mostly it is preferred for solving Classification problems. It is a tree-structured classifier, where internal nodes represent the features of a dataset, branches represent the decision rules and each leaf node represents the outcome.

In [116]:
from sklearn.tree import DecisionTreeClassifier
dtc = DecisionTreeClassifier()
dtc.fit(x_train1, y_train1)
dtc.score(x_test1, y_test1)

0.9894736842105263

In [118]:
y_pred = np.round(dtc.predict(x_test1), decimals=2)
pd.DataFrame({'Actual Marks': y_test1, 'Predicted Marks': y_pred})

Actual Marks  Predicted Marks
297            69               69
448            68               68
443            68               68
350            66               66
238            73               73
..            ...              ...
66             62               62
175            66               66
6              67               67
156            60               60
251            70               70

[95 rows x 2 columns]

### Random Forest Classifier Algorithm

Random Forest is a popular machine learning algorithm that belongs to the supervised learning technique. It can be used for both Classification and Regression problems in ML. It is based on the concept of ensemble learning, which is a process of combining multiple classifiers to solve a complex problem and to improve the performance of the model.

In [119]:
rfc = RandomForestClassifier()
rfc.fit(x_train1, y_train1)
rfc.score(x_test1, y_test1)

0.9894736842105263

### K-Nearest Neighbours Algorithm

K-Nearest Neighbour is one of the simplest Machine Learning algorithms based on Supervised Learning technique. K-NN algorithm assumes the similarity between the new case/data and available cases and put the new case into the category that is most similar to the available categories. K-NN algorithm stores all the available data and classifies a new data point based on the similarity. This means when new data appears then it can be easily classified into a well suite category by using K- NN algorithm.

In [120]:
from sklearn.neighbors import KNeighborsClassifier  
classifier= KNeighborsClassifier(n_neighbors=5, metric='minkowski', p=2 ) 

In [121]:
classifier.fit(x_train1, y_train1)  

KNeighborsClassifier()

In [122]:
classifier.score(x_test1, y_test1)

0.5052631578947369

### Gausian NB Algorithm

This method is expected to be called several times consecutively on different chunks of a dataset so as to implement out-of-core or online learning. This is especially useful when the whole dataset is too big to fit in memory at once. This method has some performance and numerical stability overhead, hence it is better to call partial_fit on chunks of data that are as large as possible (as long as fitting in the memory budget) to hide the overhead.

In [123]:
clf = GaussianNB()
clf.fit(x_train1, y_train1)

GaussianNB()

In [124]:
clf.score(x_test1, y_test1)

0.21052631578947367

### Support Vector Machine Algorithm

Support Vector Machine or SVM is one of the most popular Supervised Learning algorithms, which is used for Classification as well as Regression problems. However, primarily, it is used for Classification problems in Machine Learning. The goal of the SVM algorithm is to create the best line or decision boundary that can segregate n-dimensional space into classes so that we can easily put the new data point in the correct category in the future. This best decision boundary is called a hyperplane.

In [125]:
svm = SVC()
svm.fit(x_train1, y_train1)

SVC()

In [126]:
svm.score(x_test1, y_test1)

0.3473684210526316

### Gradient Boosting
Gradient boosting is a machine learning technique for regression, classification and other tasks, which produces a prediction model in the form of an ensemble of weak prediction models, typically decision trees.

In [127]:
from sklearn.ensemble import GradientBoostingClassifier
clf = GradientBoostingClassifier(n_estimators=300, learning_rate=1.0,max_depth=1, random_state=0)

In [128]:
clf.fit(x_train1, y_train1)

GradientBoostingClassifier(learning_rate=1.0, max_depth=1, n_estimators=300,
                           random_state=0)

In [129]:
clf.score(x_test1, y_test1)

0.08421052631578947

### MLP Classifier
Multi-layer perceptrons (MLP) make powerful classifiers that may provide superior performance compared with other classifiers, but are often criticized for the number of free parameters. Parameter selection for optimal performance is performed using measures that correlate well with generalisation error.

In [130]:
from sklearn.neural_network import MLPClassifier
clf = MLPClassifier(random_state=1, max_iter=1000)

In [131]:
clf.fit(x_train1, y_train1)

MLPClassifier(max_iter=1000, random_state=1)

In [132]:
clf.score(x_test1, y_test1)

0.2631578947368421

### Stochastic Gradient Descent (SGD)
Stochastic Gradient Descent (SGD) is a simple yet efficient optimization algorithm used to find the values of parameters/coefficients of functions that minimize a cost function. In other words, it is used for discriminative learning of linear classifiers under convex loss functions such as SVM and Logistic regression.

In [133]:
from sklearn.linear_model import SGDClassifier
sdg = SGDClassifier()

In [134]:
sdg.fit(x_train1, y_train1)

SGDClassifier()

In [135]:
sdg.score(x_test1, y_test1)

0.14736842105263157

### Linear Discriminant Analysis (LDA)
Linear Discriminant Analysis (LDA) is a dimensionality reduction technique. As the name implies dimensionality reduction techniques reduce the number of dimensions (i.e. variables) in a dataset while retaining as much information as possible.

In [136]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
lda = LinearDiscriminantAnalysis()

In [137]:
lda.fit(x_train1, y_train1)

LinearDiscriminantAnalysis()

In [138]:
lda.score(x_test1, y_test1)

0.23157894736842105

-----------------------------------

### Model Comparison
We have deployed ten machine learning algorithms and every algorithm is deployed successfully without any hesitation. We have checked the accuracy of the models based on the accuracy score of each of the models. Now let's take a look at the scores of each models.

|Name of the Model|Accuracy Score|
|:---:|:---:|
|Logistic Regression|0.23|
|Decision Tree Classifier|0.99|
|Random Forest Classifier|0.99|
|Gausian NB Algorithm|0.21|
|KNN Algorithm|0.61|
|Support Vector Machine Algorithm|0.35|
|Gradient Boosting|0.08|
|MLP Classifier|0.26|
|Stochastic Gradient Descent|0.15|
|Linear Discriminant Analysis|0.23|


### Conclusion
**Comparing all those scores scored by the machine learning algorithms, it is clear that Random Forest and Decision Tree Classifier are having the upper hand in case of this dataset and after this, we can use KNN algorithm, which is also having good score as compared to the other deployed algorithms**

Best Fitted Models ranking - 
1. Random forest classifier
2. Decision tree classifier
3. KNN
4. Support vector machine
5. MLP classifier
6. LDA
7. Logistic regression
8. Gaussian Naive Bayes
9. SGD
10. Gradient Boosting

Hooray!! The models are deployed successfully!

## Hope this project will help you! Thank you!